### PR TITLE
Add experimental support for FTS indexes in HPOS. 

### LINKED ISSUE #45721: [Enhancement]: Improve Order Search Performance When HPOS is Enabled
### Describe the solution you'd like

On sites with a large number of orders (e.g. 150k), the WooCommerce order search becomes considerably slower when HPOS is enabled and the search filter is in its default state of "All". This was a disappointment, as the expectation after enabling HPOS was that it would improve the speed of managing lots of orders in the WooCommerce admin. The cause of this problem is that the default "All" state of the search filter is an extremely expensive database query on sites with lots of orders.

This is noted in the following two tickets:
https://github.com/woocommerce/woocommerce/pull/43865#issuecomment-1988763021
https://github.com/woocommerce/woocommerce/issues/41317#issuecomment-1975186791

I would propose that we make the default state of the WooCommerce order search filter an option that is faster than "All". Perhaps one that searches similar order data as it did without HPOS enabled (when WooCommerce uses the default post meta tables)?

### Describe alternatives you've considered

A solution that might be easier to implement in the meantime would be a WordPress hook to allow a store owner to specify the default state of their store's order search filter. This would permit small stores to keep using "All" as their default but could allow larger stores to specify a different faster default order search filter (e.g. Customer).

Perhaps the best long-term solution would be a search that used a highly effective index to speed up the search query.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by vedanshujain**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

This PR adds two changes:

1. For order search, we changed the query pattern that we used in HPOS to use `SELECT....JOIN target_table...WHERE target_table.columne = ...`, instead of `SELECT ...WHERE id in ( SELECT order_id FROM target_table WHERE column = ... )` .  We did this because the query planner would suggest that the former is faster, however, based on reports we are getting, it looks like the second format is faster in reality because of data pattern distribution. This PR reverts the pattern to the latter (we use the latter in posts as well).

2. This adds experimental support for FTS indexes. This is something that we scoped in #35150 to do for HPOS, and is one of the benefits of more normalized data. This converts search queries from complex non-indexed queries to fast index searches. Note that they are marked experimental out of caution, we intend to make them default in couple of versions.

Closes #45721, #35150

Towards #41317

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

With HPOS authoritative:
First, test the query changes. (Note that there shouldn't be any functional change, only query pattern is changed to be more performant).
1. Add a few orders, using different products and customer information.
3. Go to admin view (`/wp-admin/admin.php?page=wc-orders`), and search for orders using different product names and customer names used in step 1. Check that the results are as expected.

Then, test the experimental FTS support:
1. Go to WooCommerce > Settings > Advance > Features, and enable the `HPOS Full text search indexes` feature
Screenshot 2024-04-02 at 10 03 49 PM
2. Repeat the testing instructions from previous block again, to assert that search is working as expected with FTS indexes.

<br class="Apple-interchange-newline">` 

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

<details>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Comment 

</details>

### CODE REVIEW

**Comment by lkraav** [comment]
Nice :+1: looking forward to seeing some benchmarks for FTS vs current?

**Comment by vedanshujain** [comment]
> Nice 👍 looking forward to seeing some benchmarks for FTS vs current?r

I am planning to post them soon enough, but it looks super promising 🥇

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php [hunk 1]
```diff
@@ -3,6 +3,7 @@
 namespace Automattic\WooCommerce\Internal\DataStores\Orders;
 
 use Exception;
+use phpDocumentor\Reflection\Types\This;
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php]: **This** seems to have been added by mistake (?)

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php [hunk 2]
```diff
@@ -235,6 +224,74 @@ private function generate_where_for_search_filter( string $search_filter ) : str
...
+
+		$fts_enabled = get_option( CustomOrdersTableController::HPOS_FTS_INDEX_OPTION ) === 'yes' && get_option( CustomOrdersTableController::HPOS_FTS_ADDRESS_INDEX_CREATED_OPTION ) === 'yes';
+
+		if ( $fts_enabled ) {
+			// phpcs:disable WordPress.DB.PreparedSQL.InterpolatedNotPrepared -- $order_table and $address_table are hardcoded.
+			return $wpdb->prepare(
+				"
+$order_table.id IN (
+	SELECT order_id FROM $address_table WHERE
+	MATCH( $address_table.first_name, $address_table.last_name, $address_table.address_1, $address_table.address_2, $address_table.city, $address_table.postcode, $address_table.country, $address_table.state ) AGAINST ( %s IN BOOLEAN MODE )
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php]: This doesn't seem to work correctly. When we create the full-text index on this table, we do it for the following columns: `(first_name, last_name, company, address_1, address_2, city, state, postcode, country, email)` and I believe that's the same set of columns we should be querying here. Otherwise the query fails with `Can't find FULLTEXT index matching the column list`.

Regardless of the error, we definitely want to query e-mail and company too, right?

File: plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php [hunk 3]
```diff
@@ -46,6 +47,12 @@ class CustomOrdersTableController {
 
 	public const DEFAULT_DB_TRANSACTIONS_ISOLATION_LEVEL = 'READ UNCOMMITTED';
 
+	public const HPOS_FTS_INDEX_OPTION = 'woocommerce_hpos_fts_index_enabled';
+
+	public const HPOS_FTS_ADDRESS_INDEX_CREATED_OPTION = 'woocommerce_hpos_address_fts_index_created';
+
+	public const HPOS_FTS_ORDER_ITEM_INDEX_CREATED_OPTION = 'woocommerce_hpos_order_item_fts_index_created';
```
Comment (jorgeatorres) [inline plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:55]: Is there any reason to truly have these as separate options from the feature option itself? I mean, if instead of hooking onto `updated_option` we hooked onto `pre_update_option` we could attempt to create the indexes and return `no` if we were not able to do so. That way, if the feature is enabled we can assume the indexes are there. In any case, when the feature is enabled we use `DatabaseUtil` to find whether the indexes exist, so those options are only useful at query time (but then, we could just check the feature option instead).

I'm just in favor of option economy, but do let me know if I'm missing something.

If we later add new indexes as part of this feature, we could handle those with migration routines as usual. wdyt?

**Review by jorgeatorres — CHANGES_REQUESTED** [review]
Hi @vedanshujain!

This is all super exciting. I left some comments but nothing super critical. Let me know what you think.

File: plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php (hunk 3 above)
Comment (vedanshujain) [reply plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:55]: So the utility of two different options is that I am trying to see if there are failures associated with creating this index on a possibly large table (order items). If there is, then we can evaluate a different approach or drop this one. The search query also uses these options use either the normal or the fts query, per index.

I agree with your point about options economy, but let's remove it when we promote this feature to stable (and know that both indexes are equally robust). wdyt?

File: plugins/woocommerce/includes/class-wc-install.php [hunk 4]
```diff
@@ -1595,7 +1595,8 @@ private static function get_schema() {
   KEY `stock_status` (`stock_status`),
   KEY `stock_quantity` (`stock_quantity`),
   KEY `onsale` (`onsale`),
-  KEY min_max_price (`min_price`, `max_price`)
+  KEY min_max_price (`min_price`, `max_price`),
```
Comment (vedanshujain) [inline plugins/woocommerce/includes/class-wc-install.php]: Hmm this is also accidental, but wanted. Moving this to a different PR as well.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php (hunk 1 above)
Comment (vedanshujain) [reply plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php]: Removed! thanks for catching.

File: plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php (hunk 2 above)
Comment (vedanshujain) [reply plugins/woocommerce/src/Internal/DataStores/Orders/OrdersTableSearchQuery.php]: Yes, I think I was testing on a manually created index. Thanks for catching, addressed in https://github.com/woocommerce/woocommerce/pull/46130/commits/31d4b2f06490723ffaeaf6fd5879995db0641756 plus added unit tests.

**Comment by vedanshujain** [comment]
@jorgeatorres Ready for another review!

**Comment by vedanshujain** [comment]
looking at ci failures

File: plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php (hunk 3 above)
Comment (jorgeatorres) [reply plugins/woocommerce/src/Internal/DataStores/Orders/CustomOrdersTableController.php:55]: Sounds good!

**Review by jorgeatorres — APPROVED** [review]
LGTM. Thank you @vedanshujain! Nice job!

**Comment by OllieJones** [comment]
With respect and gratitude for the hard work people are doing here, I don't think it's going to be a usable performance improvement for the `column LIKE '%searchterm%'` comprehensive grep-style search you guys presently have for orders.

FULLTEXT has limitations, like its inability to deal with less than a minimum number of characters in words. It also has stoplists, handles numbers strangely, etc. A merchant wants to be able to search orders by stuff like postcode fragment or telephone number, not just by whole words. FULLTEXT isn't made for that. (What it IS made for is a question for another day.)

I put together a solution to the order search slowness problem that builds and uses a [trigram table](https://www.plumislandmedia.net/wordpress-plugins/fast-woo-order-lookup/#how-does-it-work-trigrams). It really does speed up the search, and it has exactly the same functionality as the current order search. It's open source, so copy it as you will. https://wordpress.org/plugins/fast-woo-order-lookup/

**Comment by vedanshujain** [comment]
@OllieJones Thanks for the great feedback, my understanding (which could be wrong) is that it would work well for partial matches too, not just words, well unless you are searching for just one or two characters. It's a bit quirky in the sense that it can search for partial prefixes but not suffixes and so on, so postcode fragment search will (should?) work, at least if it's a prefix. But probably most searches are like that or at least this is how it works for many large search systems anyway.

Any case, we intend to find out how effective this is, and what parameters to fine-tune (maybe the minimum token length needs to be reduced from default 4). Maybe the stop-word list needs to be manipulated as well. This is why it's currently opt-in and marked experimental rather than being enabled by default.

> I put together a solution to the order search slowness problem that builds and uses a [trigram table](https://www.plumislandmedia.net/wordpress-plugins/fast-woo-order-lookup/#how-does-it-work-trigrams).

Note that FTS indexes also support using n-gram (which can be configured to 3 to make a trigram), although currently, we are not building the index using it. However, we can explore it if the default implementation is not effective enough. wdyt?

**Comment by OllieJones** [comment]
I did some experimenting with FULLTEXT for this sort of thing. 

I found that the FULLTEXT configuration stuff (token length, stopwords, n-gram, all that) are MariaDB  / MySQL server configuration parameters.  Most of our end-users for WordPress don't have either the access or the skills to modify those parameters. (It might be nice if the hosting team recommended particular settings, but still...)

At any rate, anything that helps these big store operators serve their customers more efficiently is good.
